### PR TITLE
Normalize terms in the language

### PR DESCRIPTION
**Pull request by Tomatosoup97**
+ De Morgan terms in the language
+ Node walker for negating the formula
+ Stringify method for the AST

### CODE REVIEW

File: bamboolean/walkers/normalize.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    flip_op = {
+        ('EQ', '=='): ('NE', '!='),
+        ('LT', '<'): ('GTE', '>='),
+        ('LTE', '<='): ('GT', '>'),
+        ('OR', 'OR'): ('AND', 'AND')
+    }
+    return {**flip_op, **flip_dict(flip_op)}[op]
+
+
+class NegateExpr(NodeVisitor):
```
Comment (mlazowik) [inline bamboolean/walkers/normalize.py]: Using verbs for classes reads weird IMO. ExprNegator?

File: bamboolean/walkers/normalize.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    def visit_Bool(self, node: Bool) -> Bool:
+        flipped_token = Token(node.token.type, not node.value)
+        return Bool(token=flipped_token)
+
+    def generic_visit(self, node: AST) -> AST:
+        return node
+
+
+class NormalizeExpr(NodeVisitor):
```
Comment (mlazowik) [inline bamboolean/walkers/normalize.py]: ditto

File: bamboolean/walkers/normalize.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+def flip_dict(d: Dict[T1, T2]) -> Dict[T2, T1]:
+    return {v: k for k, v in d.items()}
+
+
+def complementary_op(op):
+    flip_op = {
+        ('EQ', '=='): ('NE', '!='),
+        ('LT', '<'): ('GTE', '>='),
+        ('LTE', '<='): ('GT', '>'),
+        ('OR', 'OR'): ('AND', 'AND')
```
Comment (mlazowik) [inline bamboolean/walkers/normalize.py]: These being strings is kinda meh, especially as some are used in the parser and lexer too. Can you make typing & reuse stronger here?

File: bamboolean/walkers/normalize.py (hunk 1 above)
Comment (mlazowik) [inline bamboolean/walkers/normalize.py]: Using verbs for classes reads weird IMO. ExprNegator?

File: bamboolean/walkers/normalize.py (hunk 2 above)
Comment (mlazowik) [inline bamboolean/walkers/normalize.py]: ditto

File: bamboolean/tests/test_normalize.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import unittest
+
+from bamboolean.factories import normalize
+from . import fixtures
+
+
+class NormalizeExpr(unittest.TestCase):
+    def test_normalize_should_not_change_normalized_terms(self):
```
Comment (mlazowik) [inline bamboolean/tests/test_normalize.py]: It's nice to have the testing code be as minimal as is reasonable. E.g. in this case if you inline all the values not only the chance of having a bad test is lower, but the test is easier to read.

What is the gain of using a shared fixture?

File: bamboolean/walkers/normalize.py (hunk 3 above)
Comment (Tomatosoup97) [reply bamboolean/walkers/normalize.py]: Agreed, done

File: bamboolean/walkers/normalize.py (hunk 1 above)
Comment (Tomatosoup97) [reply bamboolean/walkers/normalize.py]: Done

File: bamboolean/tests/test_normalize.py (hunk 4 above)
Comment (Tomatosoup97) [reply bamboolean/tests/test_normalize.py]: Yeaaah I guess I've just used them without really thinking about this. Done

**Pull request merged**